workspaces/adventure-pack/goodies/typescript/Iterator.prototype.reduce/index.ts

```diff
@@ -0,0 +1,40 @@
+import "../Iterator.prototype.toIterable";
+import { iteratorPrototype } from "../Iterator.prototype";
+
+declare global {
+  interface Iterator<T> {
+    reduce<U>(
+      this: Iterator<T>,
+      callbackFn: (accumulator: U, value: T, index: number) => U,
+      initialValue?: U,
+    ): U;
+  }
+}
+
+iteratorPrototype.reduce ??= function <T extends U, U>(
+  this: Iterator<T>,
+  callbackFn: (accumulator: U, value: T, index: number) => T,
+  initialValue?: U,
+): U {
+  const iterator = this.toIterable()[Symbol.iterator]();
+  let index = 0;
+  let accumulator: U;
+
+  if (initialValue === undefined) {
+    const firstResult = iterator.next();
+    if (firstResult.done) {
+      throw new TypeError("Reduce of empty iterator with no initial value");
+    }
+    accumulator = firstResult.value;
+    ++index;
+  } else {
+    accumulator = initialValue;
+  }
+
+  for (let res = iterator.next(); !res.done; res = iterator.next()) {
```

Comment:
So going through `.next()` can work, but it can also be somewhat indirect. What does it look like with a regular `for..of` loop over the iterator? From the other functions implementations it seems like that might be slightly simpler?

Reply:
ack!